src/main/scala/piconot/external/Interpreter.scala

```diff
@@ -0,0 +1,133 @@
+package piconot
+
+import picolib.maze.Maze
+import picolib.semantics._
+import scala.collection.mutable._
+
+package object external {
+	def eval(program: Program): List[picolib.semantics.Rule] = {
+	    val start = Rule(Name("@"), List( SubRule(Condition(), Action(next = Next(program.start)) )))
+	    val program1 = Program(program.start, start::program.execs)
+		val program2 = taskNametoStart(program1)
+		
+		val program3 = ruleNametoTask(program2)
+		
+		val ruleList = program3.execs.map({
+			case t: Task => t.rules map {r => ruleToPicoRules(t.name.value+".", r)}
+			case r: Rule => List(ruleToPicoRules("", r))
+
+	    })
+	    
+	    ruleList.flatten.flatten
+
+	}
+
+	
+	def ruleToPicoRules(taskName: String, rule: Rule): List[picolib.semantics.Rule] = {
+	    rule.subRules map {sr => 
+	        picolib.semantics.Rule(State(taskName+rule.name.value),
+	                               conditionToSurroundings(sr.condition),
+	                               actionToMoveDir(sr.action),
+	                               (if (sr.action.next.name.value == "")
+	                                    State(taskName+rule.name.value)
+	                                else
+	                                    State(sr.action.next.name.value))
+	                              )}
+	}
+	
+	def actionToMoveDir(action: Action): MoveDirection = {
+	    action.move.direction match {
+	        case x if x eq None  => StayHere
+	        case x if x eq Up    => North
+	        case x if x eq Down  => South
+	        case x if x eq Left  => West
+	        case x if x eq Right => East
+	    }
+	}
+	
+	def conditionToSurroundings(cond: Condition): Surroundings = {
+	    var directions: MutableList[RelativeDescription] = MutableList(Anything, Anything, Anything, Anything)
+	    cond.walls.directions.foreach { d => 
+	        if (d eq Up)
+	            directions(0) = Blocked
+	        else if (d eq Right)
+	            directions(1) = Blocked
+	        else if (d eq Left)
+	            directions(2) = Blocked
+	        else if (d eq Down)
+	            directions(3) = Blocked
+	    }
+	    
+	    cond.empties.directions.foreach { d =>
+	        if (d eq Up)
+	            directions(0) = Open
+	        else if (d eq Right)
+	            directions(1) = Open
+	        else if (d eq Left)
+	            directions(2) = Open
+	        else if (d eq Down)
+	            directions(3) = Open
+	    }
+	    
+	    Surroundings(directions(0), directions(1), directions(2), directions(3))
```

Comment:
Interesting, that's really close to what we did for directions. Instead of a `foreach` we used a more recursive approach but had a very similar structure in translating our surroundings to the picobot `Surroundings`.